Fix DFG Int52 handling around negative zero

danlliu opened the pull request:
#### ea3cfb65e1ab1b5093178c62b5293ef185f3acd2
<pre>
Fix DFG Int52 handling around negative zero
<a href="https://bugs.webkit.org/show_bug.cgi?id=290851">https://bugs.webkit.org/show_bug.cgi?id=290851</a>
<a href="https://rdar.apple.com/147481105">rdar://147481105</a>

Reviewed by Yusuke Suzuki.

292253@main introduced some changes to Int52&apos;s handling of negative zero. However,
we have some cases where the AbstractInterpreter isn&apos;t tracking the possibility of
a negative zero correctly, and thus may wrongly mark states as inconsistent during
DFG/FTL compilation. This patch adds extra handling to cover cases when a negative
zero may appear.

* JSTests/stress/fiat-negative-zero-from-variable.js: Added.
(func_1_):
(func_2_):
* JSTests/stress/fiat-negative-zero.js: Added.
* Source/JavaScriptCore/dfg/DFGAbstractValue.cpp:
(JSC::DFG::AbstractValue::fixTypeForRepresentation):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupChecksInBlock):
* Source/JavaScriptCore/dfg/DFGMayExit.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
(JSC::DFG::SpeculativeJIT::convertAnyInt):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileInt52Rep):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):

Canonical link: <a href="https://commits.webkit.org/294024@main">https://commits.webkit.org/294024@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2a3d1b13cbb03e19fa7920f6f270fd36787c57c1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100583 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20235 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10534 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105720 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51171 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20543 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28709 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76594 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33639 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103590 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15760 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90853 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56950 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15575 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8858 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; Ignored 2 pre-existing failure based on results-db; Running upload-test-results") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50547 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/93247 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85489 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8935 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108074 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/99191 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27701 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20343 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85549 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28064 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87053 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85087 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21652 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29780 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7507 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21689 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27636 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32887 "Found 32 new failures in platform/audio/PlatformMediaSessionInterface.h, platform/audio/PlatformMediaSession.cpp") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/122817 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27447 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34253 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30765 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29005 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
```diff
@@ -3537,6 +3537,8 @@ void SpeculativeJIT::compile(Node* node)
             branchConvertDoubleToInt52(valueFPR, resultGPR, failureCases, scratch1GPR, scratch2FPR, canIgnoreNegativeZero);
 
+            if (canIgnoreNegativeZero)
+                node->child1().setUseKind(DoubleRepRealUse);
```
We should not change the useKind edge in code generator. This is only allowed in a certain places including Constant Folding / Strength Reduction etc.

justinmichaud replied:
Can we add an assertion for that?

Constellation replied:
Personally I don't like that since it makes all code much more complicated.

hyjorc1 commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
```diff
@@ -6611,7 +6642,12 @@ void SpeculativeJIT::convertAnyInt(Edge valueEdge, GPRReg resultGPR, bool canIgn
+    auto filter = SpecInt32Only | SpecAnyIntAsDouble;
+    if (canIgnoreNegativeZero)
+        filter |= SpecDoubleReal;
+    m_interpreter.filter(valueEdge, filter);
+    speculationCheck(BadType, JSValueRegs(valueGPR), valueEdge.node(), failureCases);
```
Can we do this instead?
```
auto speculatedType = SpecInt32Only | SpecAnyIntAsDouble;
if (canIgnoreNegativeZero)
    speculatedType |= SpecDoubleReal;
DFG_TYPE_CHECK(JSValueRegs(valueGPR), valueEdge, speculatedType, failureCases);
```

danlliu replied:
We can't quite do this, because `DFG_TYPE_CHECK` only links failure cases if the speculated type indicates that we need to filter. Here, we need to also perform range checks on the DoubleReal, so we need to ensure we link failure cases every time.

hyjorc1 commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -21936,7 +21950,10 @@ IGNORE_CLANG_WARNINGS_END
+        if (canIgnoreNegativeZero)
+            m_interpreter.filter(edge, SpecInt32Only | SpecAnyIntAsDouble | SpecDoubleReal);
+        else
+            m_interpreter.filter(edge, SpecInt32Only | SpecAnyIntAsDouble);
```
nit
```
auto speculatedType = SpecInt32Only | SpecAnyIntAsDouble;
if (canIgnoreNegativeZero)
    speculatedType |= SpecDoubleReal;
m_interpreter.filter(edge, speculatedType);
```

Constellation commented on Source/JavaScriptCore/dfg/DFGMayExit.cpp:
```diff
@@ -125,6 +124,11 @@ ExitMode mayExitImpl(Graph& graph, Node* node, StateType& state)
         break;
 
+    case Int52Rep:
+        if (bytecodeCanIgnoreNegativeZero(node->arithNodeFlags()))
```
Isn't it opposite? (If we cannot ignore negative zero, then we may exit).

Constellation replied:
I think we need to precisely model the exit here based on Edge type.
```
switch (node->child1().useKind()) {
case RealNumberUse:
    return Exits;
case AnyIntUse:
    break;
case DoubleRepAnyIntUse:
    break;
case DoubleRepRealUse:
    return Exits;
}
```

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
```diff
@@ -3543,6 +3552,28 @@ void SpeculativeJIT::compile(Node* node)
             break;
```
I think we need to put `canIgnoreNegativeZero = false` for `DoubleRepAnyIntUse` case since we need to reject `-0`.
`DoubleRepAnyIntUse` edge filter gets passed. So subsequent code will assume that input was actually meeting `DoubleRepAnyIntUse`'s values. This means that we need to exit when we see `-0` here.

Constellation commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -2255,6 +2255,12 @@ class LowerDFGToB3 {
+        }
+
         case DoubleRepAnyIntUse: {
             bool canIgnoreNegativeZero = bytecodeCanIgnoreNegativeZero(m_node->arithNodeFlags());
```
Ditto?

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
```diff
@@ -3514,14 +3514,23 @@ void SpeculativeJIT::compile(Node* node)
             GPRTemporary result(this);
             GPRReg resultGPR = result.gpr();
-            
             bool canIgnoreNegativeZero = bytecodeCanIgnoreNegativeZero(node->arithNodeFlags());
```
Ditto.

Constellation commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -2255,6 +2255,12 @@ class LowerDFGToB3 {
             return;
```
Ditto for `AnyIntUse` case too.

danlliu committed:
Fix DFG Int52 handling around negative zero
https://bugs.webkit.org/show_bug.cgi?id=290851
rdar://147481105

Reviewed by Yusuke Suzuki.

292253@main introduced some changes to Int52's handling of negative zero. However,
we have some cases where the AbstractInterpreter isn't tracking the possibility of
a negative zero correctly, and thus may wrongly mark states as inconsistent during
DFG/FTL compilation. This patch adds extra handling to cover cases when a negative
zero may appear.

* JSTests/stress/fiat-negative-zero-from-variable.js: Added.
(func_1_):
(func_2_):
* JSTests/stress/fiat-negative-zero.js: Added.
* Source/JavaScriptCore/dfg/DFGAbstractValue.cpp:
(JSC::DFG::AbstractValue::fixTypeForRepresentation):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupChecksInBlock):
* Source/JavaScriptCore/dfg/DFGMayExit.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
(JSC::DFG::SpeculativeJIT::convertAnyInt):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileInt52Rep):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):

Canonical link: https://commits.webkit.org/294024@main
